### DISCUSSION TITLE
Make the Ubuntu build with Qt6 successfully and easily

### DISCUSSION THREAD

**Discussion by tai271828** [solvcon/modmesh#53] [General]
# Update
Ubuntu 22.04 (Jammy Jellyfish) is expected to ship with qt6 so this thread should not be an issue unless we would like to support qt5 in earlier Ubuntu versions.

```
# rmadison libqt6core6
 libqt6core6 | 6.2.2+dfsg-6ubuntu2 | jammy/universe          | amd64, arm64, armhf, ppc64el, riscv64, s390x
 libqt6core6 | 6.2.4+dfsg-1ubuntu1 | jammy-proposed/universe | amd64, arm64, armhf, ppc64el, riscv64, s390x
```

## Original Description

So far building with Qt6 is not a must but a bit pesky for Ubuntu users. Is it worthy to make a better build experience for Ubuntu users? Not limited to modmesh itself, for example, add build scripts to `devenv` to include the expected Qt configuration and libs?

This topic is inspired by [this issue](https://github.com/solvcon/modmesh/issues/28#issuecomment-1087216101).

**Comment by yungyuc** [comment]
Why building with Qt is an issue on Ubuntu?

**Reply by tai271828** [reply]
Ah, I should have said "build with Qt6". I updated my description and the thread title.

Currently debian/ubuntu (and their package archive) does not ship with Qt6. Build Qt6 from source for debian/ubuntu users seem essential.

**Reply by yungyuc** [reply]
[jammy seems to have qt6 being packaged](https://packages.ubuntu.com/jammy/libqt6core6) and [it will be released later this month](https://discourse.ubuntu.com/t/jammy-jellyfish-release-schedule/23906).  That might be an option for Ubuntu users who do not want to build from source.

It might be an option to back-port the code to use Qt5.  I am not sure how easy it is, though.  I just started to learn Qt(6).

**Reply by tai271828** [reply]
Oooops! You are right. The package is in the "universe" repository and I missed to check it. My bad.

If qt6 is available in Jammy and I think it is good enough because Jammy is a LTS.

I don't think we need to back-port  to use Qt5 unless we urge to have this feature in Ubuntu 20.4 (Focal Fossa).

**Reply by yungyuc** [reply]
If you have not been able to make it run, perhaps https://github.com/solvcon/modmesh/pull/61 can help you.

**Comment by tai271828** [comment]
fwiw, Ubuntu 22.04 (Jammy) is able to build `modmesh` with Qt6 lib enabled. https://github.com/solvcon/modmesh/issues/28#issuecomment-1123985321

**Reply by yungyuc** [reply]
Can you see the viewer?

**Reply by tai271828** [reply]
Unfortunately, no. I did not see the viewer binary under `build` directory, thought the compiling completed successfully with `BUILD_QT=ON`. That being said, I invoked `python3 setup.py build_ext --cmake-args="-DBUILD_QT=ON -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)" --make-args="VERBOSE=1"` and got:

```
~/modmesh/build$ tree bdist.linux-x86_64/ lib.linux-x86_64-cpython-38/
bdist.linux-x86_64/
lib.linux-x86_64-cpython-38/
└── modmesh
    ├── __init__.py
    └── _modmesh.cpython-38-x86_64-linux-gnu.so

1 directory, 2 files
```

**Reply by yungyuc** [reply]
Please use `cmake` to build and try again.  Do not use `setup.py`.  The problem is that `setup.py` can only build Python extension module.  However, the viewer is an application embeds Python interpreter.  It is not a Python extension module.

You can find how to use `cmake` in the root `Makefile`.  `setup.py` was only created to make package manager happy with the Python extension module part.

**Reply by tai271828** [reply]
`make viewer` works. Now `./modmesh/cpp/binary/viewer/viewer` will show up!

**Reply by tai271828** [reply]
The `viewer` binary is launched. However, when clicking `run`, I got

```
$ ./modmesh/cpp/binary/viewer/viewer
Loading _modmesh_view ... succeeds
Loading modmesh ... succeeds
AttributeError: module 'modmesh' has no attribute 'StaticMesh'

At:
  <string>(6): make_2d
  <string>(29): <module>
```

I will check my build environment variable to figure if it is a bug or my environment issue.

**Reply by yungyuc** [reply]
This seems to be a symptom that you only build the `viewer` target but not the `buildext`.  The viewer needs both to work.  This may use an enhancement in the embedded Python code, but at the time being building both targets works around it.

**Reply by tai271828** [reply]
Thanks for the tip @yungyuc . `make buildext` does help. Now the viewer and REPL work for me ...via ssh X forwarding! I think my next step will be finding out a local machine shipped with Qt6 lib to make my development environment friendly if I would like to participate the viewer development further.

Nevertheless, I would like to prioritize my task to be 1) https://github.com/solvcon/modmesh/issues/75 and then 2) https://github.com/solvcon/modmesh/issues/28 first. Let me back to the Qt6 lib in my Linux distro issue later.

**Comment by tai271828** [comment]
Jim told us that we can download qt6 deb here (by filling your contact information) https://www.qt.io/download
